// waveform-playlist/lib/annotation/AnnotationList.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _h = require('virtual-dom/h');

var _h2 = _interopRequireDefault(_h);

var _aeneas = require('./input/aeneas');

var _aeneas2 = _interopRequireDefault(_aeneas);

var _aeneas3 = require('./output/aeneas');

var _aeneas4 = _interopRequireDefault(_aeneas3);

var _conversions = require('../utils/conversions');

var _DragInteraction = require('../interaction/DragInteraction');

var _DragInteraction2 = _interopRequireDefault(_DragInteraction);

var _ScrollTopHook = require('./render/ScrollTopHook');

var _ScrollTopHook2 = _interopRequireDefault(_ScrollTopHook);

var _timeformat = require('../utils/timeformat');

var _timeformat2 = _interopRequireDefault(_timeformat);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var AnnotationList = function () {
  function AnnotationList(playlist, annotations) {
    _classCallCheck(this, AnnotationList);

    this.playlist = playlist;
    this.annotations = annotations.map(function (a, i) {
      // TODO support different formats later on.
      var note = (0, _aeneas2.default)(a);
      note.leftShift = new _DragInteraction2.default(playlist, {
        direction: 'left',
        index: i
      });
      note.rightShift = new _DragInteraction2.default(playlist, {
        direction: 'right',
        index: i
      });

      return note;
    });
    this.setupEE(playlist.ee);

    // TODO actually make a real plugin system that's not terrible.
    this.playlist.isContinuousPlay = false;
    this.playlist.linkEndpoints = false;
    this.length = this.annotations.length;
  }

  _createClass(AnnotationList, [{
    key: 'setupEE',
    value: function setupEE(ee) {
      var _this = this;

      ee.on('dragged', function (deltaTime, data) {
        var annotationIndex = data.index;
        var annotations = _this.annotations;
        var note = annotations[annotationIndex];

        // resizing to the left
        if (data.direction === 'left') {
          var originalVal = note.start;
          note.start += deltaTime;

          if (note.start < 0) {
            note.start = 0;
          }

          if (annotationIndex && annotations[annotationIndex - 1].end > note.start) {
            annotations[annotationIndex - 1].end = note.start;
          }

          if (_this.playlist.linkEndpoints && annotationIndex && annotations[annotationIndex - 1].end === originalVal) {
            annotations[annotationIndex - 1].end = note.start;
          }
        } else {
          // resizing to the right
          var _originalVal = note.end;
          note.end += deltaTime;

          if (note.end > _this.playlist.duration) {
            note.end = _this.playlist.duration;
          }

          if (annotationIndex < annotations.length - 1 && annotations[annotationIndex + 1].start < note.end) {
            annotations[annotationIndex + 1].start = note.end;
          }

          if (_this.playlist.linkEndpoints && annotationIndex < annotations.length - 1 && annotations[annotationIndex + 1].start === _originalVal) {
            annotations[annotationIndex + 1].start = note.end;
          }
        }

        _this.playlist.drawRequest();
      });

      ee.on('continuousplay', function (val) {
        _this.playlist.isContinuousPlay = val;
      });

      ee.on('linkendpoints', function (val) {
        _this.playlist.linkEndpoints = val;
      });

      ee.on('annotationsrequest', function () {
        _this.export();
      });

      return ee;
    }
  }, {
    key: 'export',
    value: function _export() {
      var output = this.annotations.map(function (a) {
        return (0, _aeneas4.default)(a);
      });
      var dataStr = 'data:text/json;charset=utf-8,' + encodeURIComponent(JSON.stringify(output));
      var a = document.createElement('a');

      document.body.appendChild(a);
      a.href = dataStr;
      a.download = 'annotations.json';
      a.click();
      document.body.removeChild(a);
    }
  }, {
    key: 'render',
    value: function render() {
      var _this2 = this;

      var boxes = (0, _h2.default)('div.annotations-boxes', {
        attributes: {
          style: 'height: 30px;'
        }
      }, this.annotations.map(function (note, i) {
        var samplesPerPixel = _this2.playlist.samplesPerPixel;
        var sampleRate = _this2.playlist.sampleRate;
        var pixPerSec = sampleRate / samplesPerPixel;
        var pixOffset = (0, _conversions.secondsToPixels)(_this2.playlist.scrollLeft, samplesPerPixel, sampleRate);
        var left = Math.floor(note.start * pixPerSec - pixOffset);
        var width = Math.ceil(note.end * pixPerSec - note.start * pixPerSec);

        return (0, _h2.default)('div.annotation-box', {
          attributes: {
            style: 'position: absolute; height: 30px; width: ' + width + 'px; left: ' + left + 'px',
            'data-id': note.id
          }
        }, [AnnotationList.renderResizeLeft(note), (0, _h2.default)('span.id', {
          onclick: function onclick() {
            if (_this2.playlist.isContinuousPlay) {
              _this2.playlist.ee.emit('play', _this2.annotations[i].start);
            } else {
              _this2.playlist.ee.emit('play', _this2.annotations[i].start, _this2.annotations[i].end);
            }
          }
        }, [note.id]), AnnotationList.renderResizeRight(note)]);
      }));

      var boxesWrapper = (0, _h2.default)('div.annotations-boxes-wrapper', {
        attributes: {
          style: 'overflow: hidden;'
        }
      }, [boxes]);

      var text = (0, _h2.default)('div.annotations-text', {
        hook: new _ScrollTopHook2.default()
      }, this.annotations.map(function (note) {
        var format = (0, _timeformat2.default)(_this2.playlist.durationFormat);
        var start = format(note.start);
        var end = format(note.end);

        var segmentClass = '';
        if (_this2.playlist.isPlaying() && _this2.playlist.playbackSeconds >= note.start && _this2.playlist.playbackSeconds <= note.end) {
          segmentClass = '.current';
        }

        return (0, _h2.default)('div.row' + segmentClass, [(0, _h2.default)('span.annotation.id', [note.id]), (0, _h2.default)('span.annotation.start', [start]), (0, _h2.default)('span.annotation.end', [end]), (0, _h2.default)('span.annotation.text', [note.lines])]);
      }));

      return (0, _h2.default)('div.annotations', [boxesWrapper, text]);
    }
  }], [{
    key: 'renderResizeLeft',
    value: function renderResizeLeft(note) {
      var events = _DragInteraction2.default.getEvents();
      var config = { attributes: {
          style: 'position: absolute; height: 30px; width: 10px; top: 0; left: -2px',
          draggable: true
        } };

      events.forEach(function (event) {
        config['on' + event] = note.leftShift[event].bind(note.leftShift);
      });

      return (0, _h2.default)('div.resize-handle.resize-w', config);
    }
  }, {
    key: 'renderResizeRight',
    value: function renderResizeRight(note) {
      var events = _DragInteraction2.default.getEvents();
      var config = { attributes: {
          style: 'position: absolute; height: 30px; width: 10px; top: 0; right: -2px',
          draggable: true
        } };

      events.forEach(function (event) {
        config['on' + event] = note.rightShift[event].bind(note.rightShift);
      });

      return (0, _h2.default)('div.resize-handle.resize-e', config);
    }
  }]);

  return AnnotationList;
}();

exports.default = AnnotationList;
